### ISSUE TITLE
Support OpenShift DeploymentConfig

### ISSUE THREAD

**Issue by pradeepto** [kedgeproject/kedge#237]
Kedge should generate and deploy/undeploy/apply OpenShift `DeploymentConfig`.

**Comment by cdrage** [comment]
@pradeepto This will be solved in #250 I am going to modify my PR to implement DeploymentConfig.

**Comment by kadel** [comment]
I think that before we start implementing this in #257 we should figure out how we are going to do that.


Are we going to add fields from DeploymentConfig to Deployment?
Are we going to create new controller and leverage new controller field?
Or some other solution that?

Because each of those approaches has different cons and pros.

**Comment by kadel** [comment]
If we use new `controller` filed that it means that Kedge file will be always tied to specific distribution.
That doesn't have to be necessarily bad. As it can have different behaviour for OpenShift and Kubernetes.
We also wouldn't need `--distribution` switch as distribution could be detected from `controller` field.

If we merge DeploymentSpec with DeploymentConfigSpec it will create a mess with validation as you will have to know which distribution you are targeting. Some kedge file could be universal (in the way that they could be converted to OpenShift and also Kubernetes) but this will work only to some degree. 

It needs more research. I might be wrong and both Objects are more similar than I think (same applies for Ingress vs Route).

We should research and think about this before we jump to implementation.

**Comment by cdrage** [comment]
@kadel What I was thinking:

- Implement DeploymentConfig as a separate controller (generates both Deployment + DeploymentConfig since both are required)
- By default, when specifying `--distribution openshift` it will change the default controller of `deployment` to `deploymentconfig` as what we have now with `controller:`

**Comment by kadel** [comment]
> Implement DeploymentConfig as a separate controller (generates both Deployment + DeploymentConfig since both are required)

Why is Deployment required?



> By default, when specifying --distribution openshift it will change the default controller of deployment to deploymentconfig as what we have now with controller:

If  `--distribution` flag would only change default I would rather remove it.
Example:
You write Kedge with DeploymentConfig (without setting controller) and used some DC specific field.
Then you run kedge without `--distribution` flag, what then?

**Comment by cdrage** [comment]
@kadel Ah, you're right. Perhaps we can instead create enough docs / information that we can specify deploymentconfig as a controller, removing the --distribution flag instead.

**Comment by kadel** [comment]
I've quickly compared DeploymentSpec and DeploymentConfigSpec and it looks like the main difference is in DeploymentStrategy, otherwise base of those objects looks the same.


Kubernetes:
```go
type DeploymentStrategy struct {
	// Type of deployment. Can be "Recreate" or "RollingUpdate". Default is RollingUpdate.
	// +optional
	Type DeploymentStrategyType

	// Rolling update config params. Present only if DeploymentStrategyType =
	// RollingUpdate.
	//---
	// TODO: Update this to follow our convention for oneOf, whatever we decide it
	// to be.
	// +optional
	RollingUpdate *RollingUpdateDeployment
}
```

openshift:
```
// DeploymentStrategy describes how to perform a deployment.
type DeploymentStrategy struct {
	// Type is the name of a deployment strategy.
	Type DeploymentStrategyType `json:"type,omitempty" protobuf:"bytes,1,opt,name=type,casttype=DeploymentStrategyType"`

	// CustomParams are the input to the Custom deployment strategy, and may also
	// be specified for the Recreate and Rolling strategies to customize the execution
	// process that runs the deployment.
	CustomParams *CustomDeploymentStrategyParams `json:"customParams,omitempty" protobuf:"bytes,2,opt,name=customParams"`
	// RecreateParams are the input to the Recreate deployment strategy.
	RecreateParams *RecreateDeploymentStrategyParams `json:"recreateParams,omitempty" protobuf:"bytes,3,opt,name=recreateParams"`
	// RollingParams are the input to the Rolling deployment strategy.
	RollingParams *RollingDeploymentStrategyParams `json:"rollingParams,omitempty" protobuf:"bytes,4,opt,name=rollingParams"`

	// Resources contains resource requirements to execute the deployment and any hooks.
	Resources kapi.ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,5,opt,name=resources"`
	// Labels is a set of key, value pairs added to custom deployer and lifecycle pre/post hook pods.
	Labels map[string]string `json:"labels,omitempty" protobuf:"bytes,6,rep,name=labels"`
	// Annotations is a set of key, value pairs added to custom deployer and lifecycle pre/post hook pods.
	Annotations map[string]string `json:"annotations,omitempty" protobuf:"bytes,7,rep,name=annotations"`

	// ActiveDeadlineSeconds is the duration in seconds that the deployer pods for this deployment
	// config may be active on a node before the system actively tries to terminate them.
	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty" protobuf:"varint,8,opt,name=activeDeadlineSeconds"`
}
```

**Comment by containscafeine** [comment]
Okay, I'm a tiny bit confused here.

I'm fine with doing DeploymentConfig only with the controller field and not using `--distribution`.

However, we will have different controller structs for both, right? Right,
```golang
type DeploymentSpecMod struct {
	ControllerFields `json:",inline"`
	ext_v1beta1.DeploymentSpec `json:",inline"`
}
```
and something like -
```golang
type DeploymentConfigSpecMod struct {
	ControllerFields `json:",inline"`
	openshift.DeploymentConfig `json:",inline"`
}
```

There won't be any merging of these structs, and one won't be derived from another, etc, right?
Maybe I misread something.

**Comment by kadel** [comment]
> There won't be any merging of these structs, and one won't be derived from another, etc, right?
> Maybe I misread something.

@containscafeine yes, that is one of the options that we are currently discussing

**Comment by kadel** [comment]
Also, we shouldn't forget that adding OpenShift's DeploymentConfig  also means that we have to add ImageStreams

**Comment by surajssd** [comment]
Yes create a separate struct like @containscafeine mentioned.

And let's keep `--distribution` flag. We should not tie down the `controller` field to logic of detecting the distribution. Making these implicit ties will increase complexity and confusion for everyone(users and devs of kedge).

**Comment by kadel** [comment]
> And let's keep --distribution flag. We should not tie down the controller field to logic of detecting the distribution. Making these implicit ties will increase complexity and confusion for everyone(users and devs of kedge).

Strongly disagree here. Confusing will be requiring user to set `--distrubtion` flag. see https://github.com/kedgeproject/kedge/issues/237#issuecomment-329533190

It is not needed. Setting `controller: deploymentConfig ` is all we need to know that users want to generate for OpenShift.
Why complicate things with `--distribution`?

**Comment by kadel** [comment]
There is a small blocker. There is an [openshift/clinet-go](https://github.com/openshift/client-go) but apparently, it is not yet ready for use (I didn't try that personally)

Teoreticaly we could use [opensfhit/origin](https://github.com/openshift/origin/). The problem is that we are already using [kubernetes/client-go](https://github.com/kubernetes/client-go) and openshift/origin is [importing k8s.io/kubernetes/pkg/api](https://github.com/openshift/origin/blob/master/pkg/apps/apis/apps/v1/types.go#L8) this will create a lot of dependency and import conflicts with kubernetes/client-go :-(

**Comment by kadel** [comment]
Someone should look into this. Maybe be it will be possible to use openshift/client-go as it is. All we need is structs for DeploymentConfig and Routes nothing more

**Comment by cdrage** [comment]
New PR moved to #288

**Comment by kadel** [comment]
Issues that need to be closed before we can consider initial support for DeploymentConfigs finished:

- [ ] use oc insetad of kubectl for OpenShift artifacts #341
- [ ]  Documentation for deploymentConfig #358
- [ ] replicas set to 0 if not specified #357

**Comment by containscafeine** [comment]
All PRs merged, closing this.
